presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusClient.java

```diff
@@ -0,0 +1,164 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.plugin.prometheus;
+
+import com.facebook.airlift.json.JsonCodec;
+import com.facebook.presto.common.type.DoubleType;
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.common.type.TypeManager;
+import com.facebook.presto.spi.PrestoException;
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+
+import javax.inject.Inject;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.Files;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.facebook.presto.common.type.TimestampWithTimeZoneType.TIMESTAMP_WITH_TIME_ZONE;
+import static com.facebook.presto.common.type.VarcharType.VARCHAR;
+import static com.facebook.presto.plugin.prometheus.PrometheusColumn.mapType;
+import static com.facebook.presto.plugin.prometheus.PrometheusErrorCode.PROMETHEUS_TABLES_METRICS_RETRIEVE_ERROR;
+import static com.facebook.presto.plugin.prometheus.PrometheusErrorCode.PROMETHEUS_UNKNOWN_ERROR;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class PrometheusClient
+{
+    public static final Type TIMESTAMP_COLUMN_TYPE = TIMESTAMP_WITH_TIME_ZONE;
+    public static final String METRICS_ENDPOINT = "/api/v1/label/__name__/values";
+
+    private static final OkHttpClient httpClient = new OkHttpClient.Builder().build();
+
+    private final Optional<File> bearerTokenFile;
+    private final Supplier<Map<String, Object>> tableSupplier;
+    private final Type varcharMapType;
+
+    @Inject
+    public PrometheusClient(PrometheusConnectorConfig config, JsonCodec<Map<String, Object>> metricCodec, TypeManager typeManager)
+    {
+        requireNonNull(config, "config is null");
+        requireNonNull(metricCodec, "metricCodec is null");
+        requireNonNull(typeManager, "typeManager is null");
+
+        bearerTokenFile = config.getBearerTokenFile();
+        URI prometheusMetricsUri = getPrometheusMetricsURI(config.getPrometheusURI());
+        tableSupplier = Suppliers.memoizeWithExpiration(
+                () -> fetchMetrics(metricCodec, prometheusMetricsUri),
+                config.getCacheDuration().toMillis(),
+                MILLISECONDS);
+        varcharMapType = typeManager.getType(mapType(VARCHAR.getTypeSignature(), VARCHAR.getTypeSignature()));
```

Comment:
varcharMapType cannot be inline because it's derived from `typeManager` when create a new PrometheusClient. We need to keep a local copy of TypeManager then derive which is unncessary